doc/man3/EVP_MAC.pod

```diff
@@ -0,0 +1,338 @@
+ int EVP_MAC_ctrl_str(EVP_MAC_CTX *ctx, const char *type, const char *value);
+ int EVP_MAC_str2ctrl(EVP_MAC_CTX *ctx, int cmd, const char *value);
+ int EVP_MAC_hex2ctrl(EVP_MAC_CTX *ctx, int cmd, const char *value);
+ int EVP_MAC_nid(const EVP_MAC *mac);
+ const char *EVP_MAC_name(const EVP_MAC *mac);
+ const EVP_MAC *EVP_get_macbyname(const char *name);
+ const EVP_MAC *EVP_get_macbynid(int nid);
+ const EVP_MAC *EVP_get_macbyobj(const ASN1_OBJECT *o);
+
+=head1 DESCRIPTION
+
+These types and functions help the application to calculate MACs of
+different types and with different underlying algorithms if there are
+any.
+
+MACs are a bit complex insofar that some of them use other other
+algorithms for actual computation.  HMAC uses a digest, and CMAC uses
+a cipher.  Therefore, there are sometimes two contexts to keep track
+of, one for the MAC algorithm itself and one for the underlying
+computation algorithm if there is one.
+
+To make things less ambiguous, this manual talks about a "context" or
+"MAC context", which is to denote the MAC level context, and about a
+"underlying context", or "computation context", which is to denote the
+context for the underlying computation algorithm if there is one.
+
+=head2 Types
+
+B<EVP_MAC> is a type that holds the implementation of a MAC.
+
+B<EVP_MAC_CTX> is a context type that holds internal MAC information
+as well as a reference to a computation context, for those MACs that
+rely on an underlying computation algorithm.
+
+=head2 Context manipulation functions
+
+EVP_MAC_CTX_new() creates a new context for the MAC type C<mac>.
+EVP_MAC_CTX_new_id() creates a new context for the numerical MAC
+identity <nid>.  The created context can then be used with most other
+functions described here.
+
+EVP_MAC_CTX_free() frees the contents of the context, including an
+underlying context if there is one, as well as the context itself.
+B<NULL> is a valid parameter, for which this function is a no-op.
+
+EVP_MAC_CTX_copy() makes a deep copy of the C<src> context to the
+C<dest> context.  The C<dest> context I<must> have been created before
+calling this function.
+
+EVP_MAC_CTX_mac() returns the B<EVP_MAC> associated with the context
+C<ctx>.
+
+=head2 Computing functions
+
+EVP_MAC_init() sets up the underlying context with information given
+through diverse controls.  This should be called before calling
+EVP_MAC_update() and EVP_MAC_final().
+
+EVP_MAC_reset() resets the computation for the given context.  This
+may not be supported by the MAC implementation.
+
+EVP_MAC_update() adds C<datalen> bytes from C<data> to the MAC input.
+
+EVP_MAC_final() does the final computation and stores the result in
+the memory pointed at by C<out>, and sets its size in the B<size_t>
+the C<poutlen> points at.  If C<out> is B<NULL>, then no computation
+is made.  To figure out what the output length will be and allocate
+space for it dynamically, simply call with C<out> being B<NULL> and
```

Comment:
I'd prefer a separate interface to return the length rather than the pass NULL hack used in many places.  For API compatibility, we're need to support the latter but still.....

Reply:
You mean, like `EVP_MAC_size`?

Reply:
Either way, you make a good point.  There are already other patterns being broken here, so why not this as well...  question is how many patterns we should break in one go.

Toss a coin and decide.  I'm ambivalent and can go either way.

Reply:
I am hopeful that the KDF API will avoid the pass NULL and get the size back syndrome.  So yes, `EVP_MAC_size` sounds good to me.  I'll cope either way of course, but one API for one purpose seems more natural than overloading things.